docsrc/allblocks.nim

```diff
@@ -1,40 +1,126 @@
 import nimoji
 import nimib
+import nimpy
+import std/[math, strutils, strformat]
 
 nbInit
 
+var nbToc: NbBlock
+
+template addToc =
+  newNbBlock("nbText", false, nb, nbToc, ""):
+    nbToc.output = "## Table of Contents:\n\n"
+
+template nbCodeBlock(name:string) =
+  let anchorName = name.toLower.replace(" ", "-")
+  nbText "<a name = \"" & anchorName & "\"></a>\n### " & name & "\n\n---"
+  # see below, but any number works for a numbered list
+  nbToc.output.add "1. <a href=\"#" & anchorName & "\">" & name & "</a>\n"
+
 nbText: """
 > This nimib document provides a brief description and example for 
 > all blocks in nimib.
 
 :warning: **This document is not finished as of now.**
 """.emojize
 
-nbText: """
-### nbCodeSkip
+addToc()
 
+nbCodeBlock:"nbCodeSkip"
+nbText: """
 Similar to `nbCode`, `nbCodeSkip` is a block that displays 
 highlighted code but does not compile or run it.
 
 Example(s):
 """
 
-nbCodeSkip: 
+nbCodeSkip:
   echo "Notice how there's no output?"
   echo "The code is not compiled or executed so no output is generated!"
 
 nbCodeSkip:
   exit() # even this won't execute!
 
+nbCodeBlock:"nbCapture"
 nbText: """
-### nbCapture
 
 `nbCapture` is a block that only shows the captured output of a code block.
 
-Example(s):
+Example:
+
+The source code is pasted here to avoid you looking it up:
+```nim
+nbCapture:
+  echo "Captured!"
+```
+And below is the actual block:
 """
 
 nbCapture:
   echo "Captured!"
 
+nbCodeBlock:"nbImage"
+nbText: """
+`nbImage` enables to display your favorite pictures.
+```nim
+nbImage(url="images/todd-cravens-nimib-unsplash.jpg", caption="Blue Whale (photograph by Todd Cravens)")
+```
+Most formats are accepted. The caption is optional!
+"""
+
+nbImage(url="images/todd-cravens-nimib-unsplash.jpg", caption="Blue Whale (photograph by Todd Cravens)")
+
+nbCodeBlock:"nbTextWithCode"
+nbText:"""
+You can run Nim code directly in your markdown text with this block. It may be of use for instance for computation inside tables. See `numerical.nim`.
```

Comment:
This paragraph is incorrect. What you describe is string interpolation using the `fmt` macro which works with ordinary `nbText` equally well. `nbTextWithCode` differs from `nbText` only by the fact that the block also stores the code that produced the string passed to it. So visually and functionally there is no difference between `nbText` and `nbTextWithCode` for the average user.

Reply:
When is the code stored, reused and not reused? When you recompile your code, does it know what it can reuse code?
If there is no difference for the average user, what is the point of this block specifically?
How would you explain this block then? May I ask you to write a suggested recommendation, so I can commit it directly with Github?

Reply:
> When is the code stored, reused and not reused? When you recompile your code, does it know what it can reuse code? If there is no difference for the average user, what is the point of this block specifically?

Everything in nimib is a block, a call to `nbText` creates a new block containing the text and a call to `nbCode` creates a block containing the code and captured output. But what if you want to get the code passed to `nbText` as well and extract it? `nbText` doesn't store the code in its block (it did in the past but not anymore) so we created `nbTextWithCode` as an alternative when you want to not only get the string but also the code that created the string. The code `fmt"{1+1}"` produces the string `"2"`. `nbText` only gives us the `"2"` but `nbTextWithCode` stores the `fmt"{1+1}"` as well. The code can be accessed with `nb.blk.code` right after creating the `nbTextWithCode`. 

> How would you explain this block then? May I ask you to write a suggested recommendation, so I can commit it directly with Github?

Good question, I'll have to think about it. I'll make a suggestion tonight or tomorrow.

Reply:
```suggestion
`nbText` only stores the string it is given, but it doesn't store the code passed to `nbText`. For example, `nbText: fmt"{1+1}"` only stores the string `"2"` but not the code `fmt"{1+1}"` that produced that string. `nbTextWithCode` works like `nbText` but it also stores the code in the created block. It can be accessed with `nb.blk.code` right after the `nbTextWithCode` call. See the end of `numerical.nim` for an example.
```

Something like this, perhaps?

Reply:
I commit it but I will move it at the bottom of the block's list.
It is more subtle than other blocks and will make a good transition to a "how to create a new block section".
I might detail one example of `nb.blk.code` for the explanation to be complete after.
Thank you for the explanation.